pulp_container/app/tasks/sync_stages.py

```diff
@@ -82,6 +82,47 @@ def __init__(self, remote, signed_only):
         self.manifest_dcs = []
         self.signature_dcs = []
 
+    def _get_content_data_blocking(self, manifest):
+        saved_artifact = manifest.contentartifact_set.first().artifact
+        raw_data = saved_artifact.file
+        content_data = json.load(raw_data)
+        saved_artifact.file.close()
+        return saved_artifact, content_data, raw_data
+
+    async def _download_artifact_data(self, manifest_url, progress_report=None):
+        downloader = self.remote.get_downloader(url=manifest_url)
+        response = await downloader.run(extra_data={"headers": V2_ACCEPT_HEADERS})
+        with open(response.path, "rb") as content_file:
+            raw_data = content_file.read()
+        response.artifact_attributes["file"] = response.path
+
+        saved_artifact = await sync_to_async(_save_artifact_blocking)(response.artifact_attributes)
+        content_data = json.loads(raw_data)
+
+        if progress_report:
+            await progress_report.aincrement()
+        return saved_artifact, content_data, raw_data, response
+
+    async def _check_for_existing_manifest(self, download_tag, progress_report=None):
+        response = await download_tag
+
+        digest = response.headers.get("docker-content-digest")
+        manifest = await sync_to_async(
+            Manifest.objects.prefetch_related("contentartifact_set").filter(digest=digest).first
+        )()
+
+        if manifest:
+            saved_artifact, content_data, raw_data = await sync_to_async(
+                self._get_content_data_blocking
+            )(manifest)
+
+        else:
+            saved_artifact, content_data, raw_data, response = await self._download_artifact_data(
```

Comment:
But the `response` object on the line 120 is a result from the `else` branch of the `_download_artifact_data`. Couldn't this `response` be different from the one on line 107?

Reply:
it will be same response - the one from line 107 is a HEAD one and on line 120 is a GET one( note that you are making request here to the `response.url` form line 107 ). Same headers, just second response contains body which you don't use - later on `headers` and `url` are used only.